### PR TITLE
#2366 - ECert feedback file processing Part 2

### PR DESCRIPTION
**Pull request by dheepak-aot**
# ECert feedback file processing Part 2

## Remove the existing column 

- [x] DB Migration to remove the existing obsolete column `error_code` and drop the unique constraint based on `error_code` and add a unique constraint to use `ecert_feedback_error_id`
- [x] Removed errorCode from the entity object
- [x] Updated the IER12 code portion which was using the error code from `disbursement_feedback_error` to use error code from relationship `eCertFeedbackError`

## Update to ECert Integration

- [x] Created a service to get the key value pair of e-Cert feedback **error code** and it's respective **error id** in the table `sims.ecert_feedback_errors`. 
example: 
```json
{
  "ERR01" : 1,
  "ERR10" : 2
}
```
**Note:** Having the error code map will avoid going to database every time to get the error id for a given error code to create a `feedback error record`

- [x] Updated ECert file handler to use the error code map and create feedback entries to  `disbursement_feedback_errors` using the error id of the respective error code received. 
- [x] Updated ECert file handler to throw error and abort the processing of the file when one or more received error codes are not present in the system (in the table `sims.ecert_feedback_errors`. As per business we are suppose to get new error codes if introduced well in advance from federal.

## Upcoming PR

- Adding new column and save the file name to `sims.disbursement_feedback_errors`.
- E2E Tests.

### CODE REVIEW

File: sources/packages/backend/libs/sims-db/src/entities/disbursement-feedback-errors.model.ts [hunk 1]
```diff
@@ -26,14 +26,6 @@ export class DisbursementFeedbackErrors extends RecordDataModel {
...
   })
   dateReceived: Date;
-  /**
-   * Error Code.
-   */
-  @Column({
-    name: "error_code",
-  })
-  errorCode: string;
-  disbursementScheduleId: number;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/sims-db/src/entities/disbursement-feedback-errors.model.ts:36]: ❤️

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/ecert-full-time-feedback-integration.scheduler.ts
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/ecert-integration/ecert-full-time-feedback-integration.scheduler.ts:1]: The feedback schedulers are also good candidates for the same centralization we did for the e-Cert, right?
Not a part of this PR 😄

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts [hunk 2]
```diff
@@ -50,6 +49,7 @@ export abstract class ECertFileHandler extends ESDCFileHandler {
     private readonly eCertGenerationService: ECertGenerationService,
     private readonly disbursementScheduleErrorsService: DisbursementScheduleErrorsService,
     private readonly systemUserService: SystemUsersService,
+    private readonly eCertFeedbackErrorService: ECertFeedbackErrorService,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:52]: Not part of the PR, `disbursementScheduleService` is not used.

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts [hunk 3]
```diff
@@ -274,115 +273,229 @@ export abstract class ECertFileHandler extends ESDCFileHandler {
...
-    return processFiles;
   }
 
   /**
    * Process each individual E-Cert response file from the SFTP.
+   * @param processSummary process summary of file processing.
    * @param eCertIntegrationService
    * @param filePath E-Cert response file to be processed.
-   * @returns Process summary and errors summary.
+   * @param offeringIntensity offering intensity.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:326]: `offeringIntensity` can be removed.

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts [hunk 4]
```diff
@@ -274,115 +273,229 @@ export abstract class ECertFileHandler extends ESDCFileHandler {
...
+        ].filter(
+          (errorCode) => errorCode && !eCertFeedbackErrorCodeMap[errorCode],
+        ),
       );
-    } else {
+    if (unknownFeedbackErrorCodes.length) {
       throw new Error(
-        `${feedbackRecord.documentNumber} document number not found in disbursement_schedule table.`,
+        `The following error codes are unknown to the system. ${Array.from(
+          new Set(unknownFeedbackErrorCodes),
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:473]: If the `unknownFeedbackErrorCodes` is already a `string[]` is it not possible to use the `join` directly?

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts [hunk 5]
```diff
@@ -274,115 +273,229 @@ export abstract class ECertFileHandler extends ESDCFileHandler {
...
+          eCertFeedbackResponseRecord.errorCode5,
+        ].filter(
+          (errorCode) => errorCode && !eCertFeedbackErrorCodeMap[errorCode],
+        ),
       );
-    } else {
+    if (unknownFeedbackErrorCodes.length) {
       throw new Error(
-        `${feedbackRecord.documentNumber} document number not found in disbursement_schedule table.`,
+        `The following error codes are unknown to the system. ${Array.from(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:472]: Minor, I would suggest to have the log as below.
```
The following error codes are unknown to the system: AAA, BBB, CCC.
```

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts [hunk 6]
```diff
@@ -274,115 +273,229 @@ export abstract class ECertFileHandler extends ESDCFileHandler {
...
+        await this.createDisbursementFeedbackError(
+          recordProcessSummary,
+          eCertFeedbackResponseRecord,
+          eCertFeedbackErrorCodeMap,
+        );
+      }
+    } catch (error: unknown) {
+      // Any error caught here will abort the file processing.
+      processSummary.error(`Error processing the file ${filePath}.`, error);
+      return;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:368]: Does the return add any value here?

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts [hunk 7]
```diff
@@ -274,115 +273,229 @@ export abstract class ECertFileHandler extends ESDCFileHandler {
...
+    processSummary: ProcessSummary,
+  ) {
+    try {
+      await eCertIntegrationService.deleteFile(filePath);
+    } catch (error) {
+      // Log the error but allow the process to continue.
+      // If there was an issue only during the file removal, it will be
+      // processed again and could be deleted in the second attempt.
+      processSummary.error(
+        `Error while deleting E-Cert response file: ${filePath}`,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:497]: Missing period.

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:1]: Minor, `logger: LoggerService;` is no longer required in the service.

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts [hunk 8]
```diff
@@ -274,115 +273,229 @@ export abstract class ECertFileHandler extends ESDCFileHandler {
...
     );
-    const processFiles: ProcessSFTPResponseResult[] = [];
+    // Return if there are no files to be processed.
+    if (!filePaths.length) {
+      processSummary.info(
+        `There are no disbursement feedback error files to be processed for ${offeringIntensity}.`,
+      );
+    }
+    // Get eCert feedback error map for all the error codes.
+    let eCertFeedbackErrorCodeMap: Record<string, number>;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:298]: As a suggestion, `Record<string, number>` is used multiple times and it can have a type created just to improve the readability.
```ts
type ECertFeedbackCodeMap = Record<string, number>;
```

**Review by andrewsignori-aot — COMMENTED** [review]
Great work, most of the comments are minors or suggestions, please take a look.

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:52]: It is part of the PR. `disbursementScheduleService` was used here before which I removed.

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:473]: unknownFeedbackErrorCodes may have duplicate error codes and hence using set to get distinct error codes.

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts (hunk 6 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:368]: I just realized it is the end of the method 😊. The answer is no and I will remove it.

File: sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts [hunk 9]
```diff
@@ -274,119 +273,230 @@ export abstract class ECertFileHandler extends ESDCFileHandler {
...
       );
-    } else {
+    if (unknownFeedbackErrorCodes.length) {
       throw new Error(
-        `${feedbackRecord.documentNumber} document number not found in disbursement_schedule table.`,
+        `The following error codes are unknown to the system: ${Array.from(
+          new Set(unknownFeedbackErrorCodes),
+        )
+          .join(",")
+          .concat(".")}`,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts:474]: Minor, suggestion, can the `.concat(".")` be just a "." after the "}"?
```ts
.join(",")}.`,
```

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for making the changes, looks good 👍
